lib/clientv2/ippool.go

```diff
@@ -121,7 +169,7 @@ func (r ipPools) Watch(ctx context.Context, opts options.ListOptions) (watch.Int
 }
 
 // validateAndSetDefaults validates IPPool fields.
-func (_ ipPools) validateAndSetDefaults(pool *apiv2.IPPool) error {
+func (r ipPools) validateAndSetDefaults(ctx context.Context, pool *apiv2.IPPool, previousCIDR string) error {
```

Comment:
Might be nicer to define the function like so:

```go
func (r ipPools) validateAndSetDefaults(ctx, new, old *apiv2.IPPool) error 
```

That way we can do arbitrary validation checks on an Update() call.

Reply:
Looking at the rest of this code, my gut feeling is that it probably wants to be split up into a few different functions.

e.g.

```go
func setDefaults(pool *IPPool) {}
func validateCreate(new *IPPool) {}
func validateUpdate(new, old *IPPool) {}
```

The `validateCreate` and `validateUpdate` could still call through to commonized `validateIPPool` or something, but I think it will make the logic that only applies to one or the other a lot more clear.

Reply:
the setDefaults would be better elsewhere I agree.

the validate, less clear - it's nice to be able to return the list of validation errors in one go which is what we attempt to do.  It gets complicated if you do validateCreate and validateUpdate since there'd be a some duplication, or we'd need to have another common validation method.

Reply:
I'll ponder though.

Reply:
It's an extra function call, but it allows us to more clearly write different validation logic for Create vs Update, whereas right now we need to have a bunch of branches within a single validation function, which I find less good.

Reply:
Ok, so having actually dug a little deeper....  as it stands *most* of the validation code could be split up as Create or Update validation with a small amount of overlap.  My concern with doing that is if we relax the restrictions about changing CIDR or duplicate CIDRs etc then we'd need to undo that change.

Reply:
I'd rather change this as per your first suggestion which I like.

Reply:
We wouldn't _need_ to undo the change.

I still think it's a much better structure, even if the validation performed by each is identical. But I'm happy to let it go for now.

Reply:
ok well actually let me change as per your initial suggestion.  I actually think the whole validation side of things will need to change anyways because we want calicoctl to be able to as much validation as it can before sending in the requests - at the moment having this validation so intertwined with the Create/update processing prevents that.  So in that sense having it split out as you suggest sounds like a good idea - but I think I'd rather do that as part of a "fix validation" task which we do have coming up.